doc/source/user-profiling.rst

```diff
@@ -0,0 +1,486 @@
+function, then calls ``ray.get`` in an attempt to store the current result 
+into the list, as follows:
+
+.. code-block:: python
+
+  # This loop is suboptimal in Ray, and should only be used for the sake of this example
+  def ex1():  
+    list1 = []
+    for i in range(5):
+      list1.append(ray.get(func.remote()))
+
+For the second version **ex2**, each iteration of the loop calls the remote 
+function, and stores it into the list **without** calling ``ray.get`` each time. 
+``ray.get`` is used after the loop has finished, in preparation for processing 
+``func()``'s results:
+
+.. code-block:: python
+
+  # This loop is more proper in Ray
+  def ex2():
+    list2 = []
+    for i in range(5):
+      list2.append(func.remote())
+    ray.get(list2)
+
+Finally, as a demonstration of Ray's parallelism abilities, let's create a 
+third version **ex3** where the driver calls a second time-consuming remote 
+function in between each call to ``func()``:
+
+.. code-block:: python
+
+  # Some other time-consuming remote function
+  @ray.remote
+    def other_func():
+      time.sleep(0.2)
+
+  def ex3():
+    list3 = []
+    for i in range(5):
+      other_func.remote()
+      list2.append(func.remote())
+    ray.get(list3)
+
+
+Timing Performance Using Python's Timestamps
+--------------------------------------------
+
+One way to sanity-check the performance of the three loops is simply to
+time how long it takes to complete each loop version. We can do this using 
+python's built-in ``time`` `module`_.
+
+.. _`module`: https://docs.python.org/2/library/time.html
+
+The ``time`` module contains a useful ``time()`` function that returns the 
+current timestamp in unix time whenever it's called. We can create a generic 
+function wrapper to call ``time()`` right before and right after each loop 
+function to print out how long each loop takes overall:
+
+.. code-block:: python
+
+  # This is a generic wrapper for any driver function you want to time
+  def time_this(f):
+    def timed_wrapper(*args, **kw):
+      start_time = time.time()
+      result = f(*args, **kw)
+      end_time = time.time()
+
+      # Time taken = end_time - start_time
+      print('| func:%r args:[%r, %r] took: %2.4f seconds |' % \
+            (f.__name__, args, kw, end_time - start_time))
+      return result
+    return timed_wrapper
+
+To **always** print out how long the loop takes to run each time the loop 
+function ``ex1()`` is called, we can evoke our ``time_this`` wrapper with 
+a function decorator. This can similarly be done to functions ``ex2()``
+and ``ex3()``:
+
+.. code-block:: python
+
+  @time_this  # Added decorator
+  def ex1():
+    list1 = []
+    for i in range(5):
+      list1.append(ray.get(func.remote()))
+
+  def main():
+    ray.init()
+    ex1()
+    ex2()
+    ex3()
+
+  if __name__ == "__main__":
+    main()
+
+Alternatively, to print out the timer on **selective** calls to ``ex1()``,
```

Comment:
Why don't we get rid of this paragraph? It doesn't really add much I think and is not needed for profiling.